cpp/src/join/cross_join.cu

```diff
@@ -0,0 +1,77 @@
+#include <cudf/column/column.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/repeat.hpp>
+#include <cudf/filling.hpp>
+#include <cudf/join.hpp>
+#include <cudf/reshape.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+
+namespace cudf {
+namespace detail {
+/**
+ * @brief  Performs a cross join on two tables (left, right)
+ *
+ * The cross join returns the cartesian product of rows from each table.
+ *
+ * The approach is to repeat the left table by the number of rows in the right table
+ * and tile the right table by the number of rows in the left table.
+ *
+ * @throws cudf::logic_error if number of columns in either `left` or `right` table is 0
+ *
+ * @param[in] left                  The left table
+ * @param[in] right                 The right table
+ * @param[in] mr                    Device memory resource to use for device memory allocation
+ * @param[in] stream                Cuda stream
+ *
+ * @returns                         Result of cross joining `left` and `right` tables
+ */
+std::unique_ptr<cudf::table> cross_join(
+  cudf::table_view const& left,
+  cudf::table_view const& right,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+  cudaStream_t stream                 = 0)
+{
+  CUDF_EXPECTS(0 != left.num_columns(), "Left table is empty");
+  CUDF_EXPECTS(0 != right.num_columns(), "Right table is empty");
+
+  // If left or right table has no rows, return an empty table with all columns
+  if ((0 == left.num_rows()) || (0 == right.num_rows())) {
+    auto empty_left_columns{empty_like(left)->release()};
+    auto empty_right_columns{empty_like(right)->release()};
+    std::move(empty_right_columns.begin(),
+              empty_right_columns.end(),
+              std::back_inserter(empty_left_columns));
+    return std::make_unique<table>(std::move(empty_left_columns));
+  }
+
+  // Repeat left table
+  numeric_scalar<size_type> num_repeats{right.num_rows()};
+  auto left_repeated{detail::repeat(left, num_repeats, mr, stream)};
+
+  // Tile right table
+  auto right_tiled{cudf::tile(right, left.num_rows(), mr)};
```

Comment:
I wanted to use a `stream` argument here but it's not supported by `cudf::tile` (there is no `detail::` implementation for tile, but it uses `gather` on stream 0 internally).

Reply:
Whenever you encounter an external API that you want to use internally, you can make a detail version of it that takes streams.